compiler/sem/semexprs.nim

```diff
@@ -2094,32 +2120,53 @@ proc semProcBody(c: PContext, n: PNode): PNode =
 
   closeScope(c)
 
-proc semYieldVarResult(c: PContext, n: PNode, restype: PType) =
-  var t = skipTypes(restype, {tyGenericInst, tyAlias, tySink})
+proc semYieldVarResult(c: PContext, n: PNode, restype: PType): PNode =
+  addInNimDebugUtils(c.config, "semYieldVarResult", n, result)
+  result = n
+  let t = skipTypes(restype, {tyGenericInst, tyAlias, tySink})
   case t.kind
   of tyVar, tyLent:
     t.flags.incl tfVarIsPtr # bugfix for #4048, #4910, #6892
-    if n[0].kind in {nkHiddenStdConv, nkHiddenSubConv}:
-      n[0] = n[0][1]
-    n[0] = takeImplicitAddr(c, n[0], t.kind == tyLent)
+    let unwrappedValue =
+      case result[0].kind
+      of nkHiddenStdConv, nkHiddenSubConv:
+        result[0][1]
+      else:
+        result[0]
+    result[0] = takeImplicitAddr(c, unwrappedValue, t.kind == tyLent)
   of tyTuple:
+    var hasError = false
     for i in 0..<t.len:
       let e = skipTypes(t[i], {tyGenericInst, tyAlias, tySink})
       if e.kind in {tyVar, tyLent}:
         e.flags.incl tfVarIsPtr # bugfix for #4048, #4910, #6892
-        let tupleConstr = if n[0].kind in {nkHiddenStdConv, nkHiddenSubConv}: n[0][1] else: n[0]
-        if tupleConstr.kind in {nkPar, nkTupleConstr}:
-          if tupleConstr[i].kind == nkExprColonExpr:
+        let tupleConstr = 
+          case result[0].kind
+          of nkHiddenStdConv, nkHiddenSubConv:
+            result[0][1]
+          else:
+            result[0]
+
+        case tupleConstr.kind
+        of nkPar, nkTupleConstr:
+          case tupleConstr[i].kind
+          of nkExprColonExpr:
             tupleConstr[i][1] = takeImplicitAddr(c, tupleConstr[i][1], e.kind == tyLent)
           else:
             tupleConstr[i] = takeImplicitAddr(c, tupleConstr[i], e.kind == tyLent)
         else:
-          semReportIllformedAst(c.config, n[0], {nkPar, nkTupleConstr})
+          hasError = true
+          tupleConstr[i] = c.config.newError(result,
+            reportAst(rsemIllformedAst, result,
+                      createSemIllformedAstMsg(n, {nkPar, nkTupleConstr})))
```

Comment:
This might create cyclic ast I think.

Reply:
Whoops, that is totally wrong, yeah it should be at the same level as the node being assigned. Nice catch

Reply:
fixed